### DISCUSSION TITLE
Save language after refresh

### DISCUSSION THREAD

**Discussion by Makxzod** [sergiodxa/remix-i18next#154] [Q&A] [answered]
When i use the i18n.changeLanguage function, the new language won't be remembered if i refresh the page. I can see my <html lang="en"> won't change either. How can i save this new language?

**Comment by sergiodxa** [comment]
You need to store it somewhere like a cookie, user's session or the DB. For cookies and user session you can pass them to RemixI18next so after you save them the getLocale function will read from there, if you store in the DB query the DB directly.

You can see an example here https://github.com/dev-xo/remix-i18n-example

**Comment by matewka** [comment]
Hi @sergiodxa,

I've applied the changes from the repo you've suggested. Everything works just fine!

But I can't figure out one thing - looks like in order to change the language, the default behaviour is to navigate to `?lng={locale}`. I couldn't find the logic behind it in the example repo.
In my case I would like to do it differently - redirect user to a separate action, e.g. `POST /change-language` so the transition is done seamlessly, without "polluting" the URL search params. Is there a way to achieve that?

**Reply by sergiodxa** [reply, marked as answer]
You can use a cookie to save the locale on the action and then when loaders run remix-i18next can read from that cookie.

Here's another example app I built myself, you can see in the root loader I save in a cookie the locale https://github.com/sergiodxa/remix-vite-i18next/blob/0a9e99a200418f1c99dd9102d51928dda5114776/app/root.tsx#L20, you can do that in your action, in my case I do it in a loader so going to `?lng={locale}` will make the loader use that locale and save it, then if the search param is not set anymore it will still use the selected locale.
